src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
+
+    # set the progress of the future
+    total_duration = acqui_task.estimate_total_duration()
+    future.set_end_time(time.time() + total_duration)
+
+    # assign the acquisition task to the future
+    executeAsyncTask(future, acqui_task.run)
+
+    return future
+
+
+class ZStackAcquisitionTask(object):
+    """
+    This class represents an acquisition task for a zstack of data.
+    """
+
+    def __init__(self, future, streams, zlevels, settings_obs):
+        """
+        The class constructor
+        future (ProgressiveFuture): the future that will execute the task asynchronously
+        streams (list of Stream): the streams to be acquired
+        zlevels (dict Stream -> list of floats): a dictionary containing the streams and the
+        corresponding lists of actuator z positions
+        settings_obs (SettingsObserver):
+        """
+        self._main_future = future
+        self._future_state = None
+        self._streams = sortStreams(streams)
+        self._zpos = next((l for l in zlevels.values() if len(l) > 1), None)
+        self._zlevels = zlevels
+        self._settings_obs = settings_obs
+        self._zstep_duration = self._estimate_zstep_duration()
+
+    def cancel(self, _):
+        """
+        Called when the user presses the button "cancel" and the future is still running.
+        (but there might  be some delay (not instantly called) )
+        return (Boolean): True if canceled, False otherwise
+        """
+        self._future_state = CANCELLED
+        # cancel the sub-futures
+        self._actuator_f.cancel()
+        self._single_acqui_f.cancel()
+        return True
+
+    def _estimate_zstep_duration(self):
+        """
+        return (float > 0): estimated time (in s) that it takes to move the focus actuator
+          by one step.
+        """
+        # get the focuser of any of the FM streams
+        focuser = next(
+            (s.focuser for s in self._streams if isinstance(s, FluoStream)), None
+        )
+        speed = None
+        if focuser is not None:
+            if model.hasVA(focuser, "speed"):
+                speed = focuser.speed.value.get("z", None)
+        if speed is None:
+            speed = 10e-6  # m/s, pessimistic
+        zstep = 0
+        if self._zpos:
+            zstep = abs(self._zpos[0] - self._zpos[1])
+        return estimateMoveDuration(zstep, speed, 0.01)
+
+    def estimate_total_duration(self):
+        """
+        Estimates the total time for the zstack acquisition.
+        return (Float > 0): the estimated time
+        """
+        acqui_time = 0.0
+        zstep_time = 0
+        nr_FM_streams = 0
+        for s in self._streams:
+            acqui_time += s.estimateAcquisitionTime() * len(self._zlevels[s])
+            if isinstance(s, FluoStream):
+                nr_FM_streams += 1
+        if self._zpos:
+            zstep_time = self._zstep_duration * (len(self._zpos) - 1) * nr_FM_streams
+        return acqui_time + zstep_time
```

Comment:
This code is nearly duplicated in acquisition.py. I suggest you use this code for both cases, you can take inspiration from the tiled acquisition task, please have a look at `estimateTiledAcquisitionTime()` function and its usage in the OverviewAcquisitionDialog.